src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -2786,3 +2861,290 @@ def find_q(m, m4_fac, D):
+        ....:     print(E)
+        verbose 4 (2865: ell_finite_field.py, EllipticCurve_with_prime_order) Testing D=-19
+        ...
+        verbose 4 (2865: ell_finite_field.py, EllipticCurve_with_prime_order) Testing D=-163
+        verbose 2 (2865: ell_finite_field.py, EllipticCurve_with_prime_order) Computing the Hilbert class polynomial H_-163
+        Elliptic Curve defined by y^2 = x^3 + 265977778*x + 120868502 over Finite Field of size 1000041437
+        verbose 4 (2865: ell_finite_field.py, EllipticCurve_with_prime_order) Testing D=-179
+        ...
+        verbose 4 (2865: ell_finite_field.py, EllipticCurve_with_prime_order) Testing D=-667
+        verbose 2 (2865: ell_finite_field.py, EllipticCurve_with_prime_order) Computing the Hilbert class polynomial H_-667
+        Elliptic Curve defined by y^2 = x^3 + 665393686*x + 948152000 over Finite Field of size 999969307
+        Elliptic Curve defined by y^2 = x^3 + 572311614*x + 178583984 over Finite Field of size 999969307
+
+    TESTS::
+
+        sage: list(EllipticCurve_with_prime_order(2))
+        [Elliptic Curve defined by y^2 + x*y + y = x^3 + 1 over Finite Field of size 2,
+         Elliptic Curve defined by y^2 = x^3 + 2*x^2 + 2 over Finite Field of size 3,
+         Elliptic Curve defined by y^2 = x^3 + 2*x over Finite Field of size 5]
+
+        sage: set_verbose(0)
+        sage: for N in prime_range(3, 100):
+        ....:     E = next(EllipticCurve_with_prime_order(N))
+        ....:     assert E.has_order(N)
+
+        sage: N = 113
+        sage: for _, E in zip(range(30), EllipticCurve_with_prime_order(N)):
+        ....:     assert E.has_order(N)
+
+        sage: N = 15175980689839334471
+        sage: E = next(EllipticCurve_with_prime_order(N))
+        sage: E.has_order(N)
+        True
+
+        sage: N = next_prime(123456789)
+        sage: E = next(EllipticCurve_with_prime_order(N))
+        sage: E.has_order(N)
+        True
+
+        sage: N = 123456789
+        sage: E = next(EllipticCurve_with_prime_order(N))
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not a prime
+
+        sage: E = next(EllipticCurve_with_prime_order(0))
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not a prime
+
+        sage: E = next(EllipticCurve_with_prime_order(-7))
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not a prime
+
+    AUTHORS:
+
+    - Martin Grenouilloux, Gareth Ma (2024-09): initial implementation
+    """
+    import itertools
+    from sage.arith.misc import is_prime, legendre_symbol
+    from sage.misc.verbose import verbose
+    from sage.quadratic_forms.binary_qf import BinaryQF
+    from sage.rings.fast_arith import prime_range
+    from sage.schemes.elliptic_curves.cm import hilbert_class_polynomial
+    from sage.sets.primes import Primes
+
+    if not is_prime(N):
```

Comment:
Since this code relies on a few probabilistic algorithms, including `has_order` which sometimes returns the wrong answer, I think it would be worth adding an optional `proof` argument to this function and using it for this prime check and any calls to `has_order`.

Reply:
That is incorrect, please read my notes above. The `.has_order` call right now is *deterministically (100%) correct*.

Reply:
I'm confused. Doesn't it give the wrong answer for this example?

```
sage: E = EllipticCurve(GF(127), [0, 1])
sage: E.order()
108
sage: E.has_order(126)
True
```

Reply:
![image](https://github.com/user-attachments/assets/e5d4ce9e-c156-45ea-8fe1-c4cb260fd84b)

Reply:
So #38617 is unrelated? Your docstring says "This demonstrates the bug mentioned in the NOTE above. The return value should be ``False`` after :issue:`38617` is fixed" for one of your examples.

Can you either update the docstring for that to explain the error better or if able, fix the bug causing the false positive?

Reply:
I think you're confusing two pieces of code here.

One part is the `has_order` function, which is implemented for all elliptic curves. That returns false positives (`True` when real answer is `False`) sometimes, which is bug #38617. That is documented correctly in the docstrings.

However, the bug only affects curves with a specific structure. In particular, when the curve has prime order or the tested value is prime (and at least $5$ or something), the `has_order` function always return the correct value. In fact it does with `num_checks=1`.

This leads to the other part when I use `has_order`, which is in `EllipticCurve_with_prime_order`. There the tested value is prime, so as mentioned above, there is no bug, so `EllipticCurve_with_prime_order` has no bug. This is why the `proof` argument you suggested is unnecessary.